### PR TITLE
[pytree] traverse `dict` in sorted key ordering

### LINKED ISSUE #114392: [pytree] equal `dict`s do not imply equal leaves and equal treespecs
### 🐛 Describe the bug

Currently, the Python pytree does not guarantee "equal `dict`s imply equal leaves and equal treespecs".

```python
In [1]: import torch.utils._pytree as pytree

In [2]: {'a': 1, 'b': 2} == {'b': 2, 'a': 1}
Out[2]: True

In [3]: pytree.tree_leaves({'a': 1, 'b': 2})
Out[3]: [1, 2]

In [4]: pytree.tree_leaves({'b': 2, 'a': 1})
Out[4]: [2, 1]

In [5]: pytree.tree_structure({'a': 1, 'b': 2})
Out[5]: 
TreeSpec(dict, ['a', 'b'], [*,
  *])

In [6]: pytree.tree_structure({'b': 2, 'a': 1})
Out[6]: 
TreeSpec(dict, ['b', 'a'], [*,
  *])

In [7]: pytree.tree_leaves({'a': 1, 'b': 2}) == pytree.tree_leaves({'b': 2, 'a': 1})
Out[7]: False

In [8]: pytree.tree_structure({'a': 1, 'b': 2}) == pytree.tree_structure({'b': 2, 'a': 1})
Out[8]: False
```

### Versions

<details>
<summary>Versions</summary>

```text
Collecting environment information...
PyTorch version: 2.2.0.dev20231121
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.7.1 (x86_64)
GCC version: Could not collect
Clang version: 17.0.5
CMake version: version 3.27.8
Libc version: N/A

Python version: 3.11.5 (main, Sep 11 2023, 08:19:27) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.6.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.0
[pip3] optree==0.10.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2023.1.0         h8e150cf_43560  
[conda] numpy                     1.26.0                   pypi_0  
[conda] optree                    0.10.0                   pypi_0  
[conda] torchfix                  0.2.0                    pypi_0
```

</details>

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114947
* #116375
* #116372
* #116374

Fixes #114392

- #114392

Python `dict` and `defaultdict` do not take the order of keys into account while comparing two dictionaries.

```python
In [1]: from collections import *

In [2]: {'a': 1, 'b': 2} == {'b': 2, 'a': 1}
Out[2]: True

In [3]: defaultdict(int, {'a': 1, 'b': 2}) == defaultdict(int, {'b': 2, 'a': 1})
Out[3]: True

In [4]: OrderedDict({'a': 1, 'b': 2}) == OrderedDict({'b': 2, 'a': 1})
Out[4]: False
```

Before this PR, the traversing order of the `dict` and `defaultdict` nodes are in insertion order. This means if two equal `dict`s have the same keys but inserted in different order, the result leaves are different:

```python
In [5]: import torch.utils._pytree as pytree

In [6]: pytree.tree_leaves({'a': 1, 'b': 2})
Out[6]: [1, 2]

In [7]: pytree.tree_leaves({'b': 2, 'a': 1})
Out[7]: [2, 1]
```

Also we will get different `TreeSpec` objects because the context of the `TreeSpec` of a `dict` node is a list of keys in insertion order. But comparing between these two lists of keys will consider the order of elements.

```python
In [8]: spec1 = pytree.tree_structure({'a': 1, 'b': 2})

In [9]: spec2 = pytree.tree_structure({'b': 2, 'a': 1})

In [10]: spec1
Out[10]: 
TreeSpec(dict, ['a', 'b'], [*,
  *])

In [11]: spec2
Out[11]: 
TreeSpec(dict, ['b', 'a'], [*,
  *])

In [12]: spec1 == spec2
Out[12]: False

In [13]: spec1.context
Out[13]: ['a', 'b']

In [14]: spec2.context
Out[14]: ['b', 'a']

In [15]: spec1.context == spec2.context
Out[15]: False
```

Not only contexts differ, but the order of the `children_specs` is also related.

------

This PR makes the traversal order of `dict` / `defaultdict` follow the sorted key order. This makes the behavior of `dict` traversal consistent with optree and JAX pytree. It also changed the context for a dictionary:

- old context: a list of keys in insertion order
- new context: a list consists of two elements:
    1. a list of sorted keys
    2. a dictionary with the original keys in insertion order, but all values are `None`. This is used to preserve the original insertion order while doing unflattening.

Some notes of the traversal order for `dict` type:

1. PyTorch before this PR: traverse `dict` in insertion order. Preserve the key order for `unflatten(flatten(dict))`.
    - Pros:
        - It's intuitive.
        - Do not have overhead for sorting.
        - Do not require the keys to be sortable.
        - Preserve the key order for `unflatten(flatten(dict))`.
    - Cons:
        - Do not guarantee equal `dict` get equal leaves and equal treespecs. Might be bad for flattening function keyword arguments (`**kwargs`).
2. JAX pytree: traverse `dict` in sorted order. Unflatten the `dict` back in sorted order rather than the original insertion order.
    - Pros:
        - Guarantee equal `dict` get equal leaves and equal treespecs.
    - Cons:
        - It's not intuitive. Need documentation.
        - Have a non-zero overhead for sorting.
        - Require the keys to be sortable.
        - Do not preserve the key order for `unflatten(flatten(dict))`.
3. optree: traverse `dict` in sorted order. Preserve the key order for `unflatten(flatten(dict))`.
    - Pros:
        - Guarantee equal `dict` get equal leaves and equal treespecs.
        - Preserve the key order for `unflatten(flatten(dict))`.
    - Cons:
        - It's not intuitive if users only use `tree_flatten` or combine `d.values()` with `tree_flatten(d)`. No concern about `tree_map` because we will do `tree_unflatten` in it.
        - Have a non-zero overhead for sorting.

cc @zou3519 @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4 @penguinwu

### CODE REVIEW

File: torch/utils/_pytree.py [hunk 1]
```diff
@@ -283,11 +283,44 @@ def _private_register_pytree_node(
 
 
 def _dict_flatten(d: Dict[Any, Any]) -> Tuple[List[Any], Context]:
-    return list(d.values()), list(d.keys())
+    sorted_keys = sorted(d)
+    sorted_values = [d[key] for key in sorted_keys]
+    return sorted_values, [sorted_keys, dict.fromkeys(d)]
```
Comment (XuehaiPan) [inline torch/utils/_pytree.py:288]: Dynamo does not support `dict.fromkeys(iterable, value=None, /)` yet, which is a `classmethod` for `dict`.

I can temporarily replace this to:

```diff
- dict.fromkeys(iterable, value)
+ {k: value for k in iterable}

- dict.fromkeys(iterable)
+ {k: None for k in iterable}
```

But in general, we may need to implement a special path for it in `BuiltinVariable.call_function`.

The error:

```text
torch._dynamo.exc.Unsupported: call_method BuiltinVariable(dict) fromkeys [ConstDictVariable()] {}
```

File: torch/utils/_pytree.py [hunk 2]
```diff
@@ -283,11 +283,44 @@ def _private_register_pytree_node(
...
 
 
 def _dict_unflatten(values: Iterable[Any], context: Context) -> Dict[Any, Any]:
-    return dict(zip(context, values))
+    sorted_keys, original_copy = context
+    d: Dict[Any, Any] = original_copy.copy()
+    d.update(zip(sorted_keys, values))
```
Comment (XuehaiPan) [inline torch/utils/_pytree.py:294]: - #115011

File: torch/utils/_pytree.py (hunk 1 above)
Comment (Skylion007) [inline torch/utils/_pytree.py:288]: Just sort the keys and values at once by passing dict items and split the item into two lists. That way you only need to do one key lookup.

**Review by Skylion007 — COMMENTED** [review]
Comment about dict flattening

File: torch/utils/_pytree.py (hunk 1 above)
Comment (XuehaiPan) [reply torch/utils/_pytree.py:288]: Here are some benchmark results:

```python
In [1]: import string

In [2]: chars = string.ascii_letters + string.digits

In [3]: import random

In [4]: random_keys = [''.join(random.choices(chars, k=8)) for _ in range(10)]

In [5]: d = dict.fromkeys(random_keys)

In [6]: def flatten1(d):
   ...:     sorted_keys = sorted(d)
   ...:     sorted_values = [d[k] for k in sorted_keys]
   ...:     return sorted_values, (sorted_keys, dict.fromkeys(d))
   ...:

In [7]: def flatten2(d):
   ...:     if not d:
   ...:         return [], ([], {})
   ...:     sorted_keys, sorted_values = zip(*sorted(d.items(), key=lambda kv: kv[0]))
   ...:     return sorted_values, (sorted_keys, dict.fromkeys(d))
   ...:

In [8]: %timeit flatten1(d)
728 ns ± 2.51 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [9]: %timeit flatten2(d)
1.05 µs ± 3.34 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

Comment (Skylion007) [reply torch/utils/_pytree.py:289]: Pretty sure these benchmarks are a tad flawed as the dictionaries are really small (10 items). You won't see much of a difference in key access unless you increase the size of the dict by several order of magnitudes.

**Pull request closed without merging**